feat(search): support code search by zoekt

adlternative opened the pull request:
### Abstract

[Zoekt](https://github.com/sourcegraph/zoekt) is an open-source search engine specifically designed for code search, utilizing 3-gram indexing for efficient segmentation. By replacing Elasticsearch/Bleve with Zoekt, it provides Gitea with precise code search capabilities and support for regular expression searches.

### Motivation
The existing code search functionality is implemented using Elasticsearch/bleve. Although Elasticsearch/bleve excels in general search domains, its disadvantages in code search are obvious:

1. Unable to support precise match searches, for example, when punctuation marks appear in the search criteria.
2. Unable to easily support regex match searches.

### Proposal
#### Goals
Support precise substring searches
Support regex searches
#### Non-Goals
Support multi-branch searches
Support code symbol syntax searches

### Competitive Product Analysis

| **Platform**    | **Search Engine**        | **Supports Regex Search** | **Supports Full Repository Search** |
|-----------------|--------------------------|:-------------------------:|:-----------------------------------:|
| **GitHub**      | Blackbird (Proprietary)  | ✓                         | ✓                                   |
| **GitLab**      | Elasticsearch / Zoekt    | ✓                         | ✓                                   |
| **grep.app**    | Closed Source            | ✓                         | ✓                                   |
| **Sourcegraph** | Zoekt    | ✓                         | ✓                                   |
| **Gitea**(us)       | Elasticsearch or Bleve   | ✗                         | ✗                                   |


### Design

#### Index
Since Zoekt is written in Golang, its API can be directly integrated through its Go package using indexBuilder.Add() and indexBuilder.MarkFileAsChangedOrRemoved() to add or remove indexed files. The fundamental processes for implementing full and incremental repository indexing in Zoekt do not differ significantly from those in Elasticsearch (ES) or Bleve.

#### Search

We can use shards.NewDirectorySearcher() or shards.NewDirectorySearcherFast() to build a searcher for searching. The search modes will support:
* exact – Complete match of any content (including punctuation)

<img width="1344" alt="image" src="https://github.com/user-attachments/assets/e66f53b4-7ca1-498a-bc07-fe1b3160e717" />

* words – Split by spaces into multiple search conditions and perform an OR query

<img width="1299" alt="image" src="https://github.com/user-attachments/assets/e13e84be-c153-4257-908b-cf272c35060d" />

* regexp – Regular expression search

<img width="1379" alt="image" src="https://github.com/user-attachments/assets/3d37644a-f50d-4139-9126-d8eed85fc465" />


* zoekt – Using the Zoekt search syntax
<img width="1387" alt="image" src="https://github.com/user-attachments/assets/307f1c43-6bd1-45d7-b81e-23b3055f6387" />


Since the search is currently limited to a single repository, we will retrieve all the content first and then handle pagination.


#### Use Method

enable this in app.ini
```ini
[indexer]
REPO_INDEXER_TYPE = zoekt
REPO_INDEXER_ENABLED = true
REPO_INDEXER_PATH = indexers/repos.zoekt
```

#### Resource Usage

Building the index in Zoekt requires 1.2 times the corpus size in RAM, and the index storage size is about three times the corpus size. Maybe we should expose some of Zoekt's internal Prometheus metrics in the future?

### Exists Issues
- [x] (solved) zoekt cannot build on i386: https://github.com/sourcegraph/zoekt/issues/935
- [x] zoekt cannot build on windows: https://github.com/sourcegraph/zoekt/pull/909
- [ ] setting.Indexer.IncludePatterns and setting.Indexer.ExcludePatterns use glob syntax, whereas Zoekt itself only supports regular expressions (regexp). It may be necessary to refactor IncludePatterns and ExcludePatterns to support regular expressions.


Try to support #33702

wxiaoguang commented:
There are already so many search engines builtin into Gitea. Many of them have various bugs.

So the questions are:

1. Will more search engines be added into Gitea to make Gitea have plenty of builtin  search engines?
2. Will the search engines become unmaintained and the bugs will never be fixed?

hiifong commented:
To be honest I prefer this zoekt search engine compared to the existing search engine

lunny commented:
maybe this can replace bleve but we need some comparsion tests.

wxiaoguang commented:
> To be honest I prefer this zoekt search engine compared to the existing search engine

That's understandable. So a few months later, another one feels "yoekt" is better, then introduce "yoekt", then a few months later, someone feels "xoekt" is better, then introduce "yoekt", and then "woekt", "voekt", "uoekt" ... "coekt", "boekt", "aoekt". Then Gitea contains all search engines on the internet.

----

I do not mean objection to introduce improvements. But actually it needs to:

1. Clarify the existing problems & fix existing problems.
2. Remove unnecessary search engine before introducing new ones.

So a clear roadmap about the "search engine plan" is necessary.

wxiaoguang commented:
> In my opinion, supporting multiple search engines is a good thing, as users may have different needs. Even GitLab now supports both ES and Zoekt search engines. see https://docs.gitlab.com/user/search
> I'm not too worried about this; Gitea should have good community maintenance. It might be because the code search functionality is not exposed by default, so many bugs haven't been discovered.

Well, do you know how many search engines are in Gitea now? And what longstanding bugs do they have?

adlternative commented:
> > To be honest I prefer this zoekt search engine compared to the existing search engine
> 
> That's understandable. So a few months later, another one feels "yoekt" is better, then introduce "yoekt", then a few months later, someone feels "xoekt" is better, then introduce "xoekt", and then "woekt", "voekt", "uoekt" ... "coekt", "boekt", "aoekt". Then Gitea contains all search engines on the internet.
> 

you don't need to worry about this: zoekt is a popular code search engine, currently used by code platforms like Gerrit, Sourcegraph, and GitLab, wrote by Gerrit author, and maintained by Sourcegraph. Zoekt has advantages that traditional search engines (like ES) do not possess: support for regex matching, substring search, etc. I don't think any new open-source code search engines will be able to replace it in the short term.

> I do not mean objection to introduce improvements. But actually it needs to:
> 
> 1. Clarify the existing problems & fix existing problems.
> 2. Remove unnecessary search engine before introducing new ones.
> 
> So a clear roadmap about the "search engine plan" is necessary.

You are right, where should the roadmap be written? I don't have experience with this. I will supplement its documentation when the zoekt functionality is more complete

wxiaoguang commented:
> I don't think any new open-source code search engines will be able to replace it in the short term.

Yep, if zoekt wins, we need to drop some others.

adlternative commented:
> > In my opinion, supporting multiple search engines is a good thing, as users may have different needs. Even GitLab now supports both ES and Zoekt search engines. see https://docs.gitlab.com/user/search
> > I'm not too worried about this; Gitea should have good community maintenance. It might be because the code search functionality is not exposed by default, so many bugs haven't been discovered.
> 
> Well, do you know how many search engines are in Gitea now? And what longstanding bugs do they have? https://github.com/go-gitea/gitea/issues?q=is%3Aissue%20state%3Aopen%20code%20search
> 
> And some bugs didn't get fixed in months, for example: "Search Functionality Issues with Bleve Engine #31565", I don't see "good community maintenance"

Sure, it's regrettable that this part of the content is unmaintained. However, for the zoekt code search, I can commit to maintaining it thoroughly.

wxiaoguang commented:
To make the code clear, we need to refactor the related code first:  Refactor issue & code search #33860 

Each "indexer" should provide the "search modes" they support by themselves. And we need to remove the "fuzzy" search for code.

adlternative commented:
@wxiaoguang @lunny I think this CR is ready for an initial review. :)

hiifong commented:
Can you simply tell me how to enable zoekt? I would like to try it.

adlternative commented:
> Can you simply tell me how to enable zoekt? I would like to try it.

write this to `custom/conf/app.ini`
```ini
[indexer]
REPO_INDEXER_TYPE = zoekt
REPO_INDEXER_ENABLED = true
REPO_INDEXER_PATH = indexers/repos.zoekt
```

hiifong commented:
Searching with zoekt syntax, typing: `lang:go` gitea will panic.

![image](https://github.com/user-attachments/assets/45f0aa87-89f9-447c-b424-bca217ee3751)

adlternative commented:
> Searching with zoekt syntax, typing: `lang:go` gitea will panic.
> 
> ![image](https://private-user-images.githubusercontent.com/89133723/430767420-45f0aa87-89f9-447c-b424-bca217ee3751.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwOTMxOTQsIm5iZiI6MTc0NDA5Mjg5NCwicGF0aCI6Ii84OTEzMzcyMy80MzA3Njc0MjAtNDVmMGFhODctODlmOS00NDdjLWI0MjQtYmNhMjE3ZWUzNzUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDA2MTQ1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiNTgwODFhNTQ3ODMwZmU4NjY5MDc1MDM0NDk2MmFkNDFlNThjMDJmZWVlNDI3NzI4MjJjNzgwOGY2ZDU5YmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MP72-mi8zGOSmUUpX_457eqxMAEwbya8pDJcBogc8mg)

I have fix it, you can try again...

hiifong commented:
> > Searching with zoekt syntax, typing: `lang:go` gitea will panic.
> > ![image](https://private-user-images.githubusercontent.com/89133723/430767420-45f0aa87-89f9-447c-b424-bca217ee3751.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwOTMxOTQsIm5iZiI6MTc0NDA5Mjg5NCwicGF0aCI6Ii84OTEzMzcyMy80MzA3Njc0MjAtNDVmMGFhODctODlmOS00NDdjLWI0MjQtYmNhMjE3ZWUzNzUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDA2MTQ1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiNTgwODFhNTQ3ODMwZmU4NjY5MDc1MDM0NDk2MmFkNDFlNThjMDJmZWVlNDI3NzI4MjJjNzgwOGY2ZDU5YmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MP72-mi8zGOSmUUpX_457eqxMAEwbya8pDJcBogc8mg)
> 
> I have fix it, you can try again...

Thank you very much, I'll do some functional testing of the feature when I have time!

hiifong commented:
I tested it briefly and found the following problem: 

1. Branches other than the default branch cannot be searched.
![image](https://github.com/user-attachments/assets/5d9825b4-1ff3-471d-8aab-f8665d3cb3d4)

2. The search results are not highlighted, it would be nice if this was available.
![image](https://github.com/user-attachments/assets/5cbbf111-ee06-459d-9b7d-d4b09c523f58)

adlternative commented:
> I tested it briefly and found the following problem:
> 
> 1. Branches other than the default branch cannot be searched.
>    ![image](https://private-user-images.githubusercontent.com/89133723/432998209-5d9825b4-1ff3-471d-8aab-f8665d3cb3d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NjY4OTcsIm5iZiI6MTc0NDQ2NjU5NywicGF0aCI6Ii84OTEzMzcyMy80MzI5OTgyMDktNWQ5ODI1YjQtMWZmMy00NzFkLThhYWItZjg2NjVkM2NiM2Q0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEyVDE0MDMxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkNWNhOGEwZmY3MDA5NDg2MjU4NTBiOWVmOWYzZWFmYWZmMmUzYTJiZTBiODQ1MzBkZTdhOGZjYWZjMWE4ODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hbn7ZBgmNeqgGiXWRjweIBXZOP23ogjvjvMcOxUa0S4)
> 2. The search results are not highlighted, it would be nice if this was available.
>    ![image](https://private-user-images.githubusercontent.com/89133723/432998682-5cbbf111-ee06-459d-9b7d-d4b09c523f58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NjY4OTcsIm5iZiI6MTc0NDQ2NjU5NywicGF0aCI6Ii84OTEzMzcyMy80MzI5OTg2ODItNWNiYmYxMTEtZWUwNi00NTlkLTliN2QtZDRiMDljNTIzZjU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEyVDE0MDMxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViNTZlMzlmMTRjYzU2MDZiMjBlMTAzOThmMWFjMjc0OWQwNWQxYTllZWJhM2U0MTRiNDFhZTk0OGQyY2ExOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dwq9kgxj0cs093gJARWUpXdwTHMN1zFdyS84Q5LOQ6k)



> I tested it briefly and found the following problem:
> 
> 1. Branches other than the default branch cannot be searched.
>    ![image](https://private-user-images.githubusercontent.com/89133723/432998209-5d9825b4-1ff3-471d-8aab-f8665d3cb3d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NjY5NzYsIm5iZiI6MTc0NDQ2NjY3NiwicGF0aCI6Ii84OTEzMzcyMy80MzI5OTgyMDktNWQ5ODI1YjQtMWZmMy00NzFkLThhYWItZjg2NjVkM2NiM2Q0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEyVDE0MDQzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1YzUwNzYwMGVlZGZkMWJiY2JlYzFiMmU5YTc5YTZlMGM0ZTViYWE4NTZlYjFjNmRmYWUxNTVhNjkzMDg2MjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qimbObRbSLu17ZUbz-CG96AS6QxDTSMlFsHqe0a49Tw)

Yes, currently only supporting the repository's default branch is the designed behavior. 
Supporting multiple branches would require some more complex refactoring.

Currently, zoekt supports indexing up to64 branches, which means we still need to have a strategy to decide which branches to configure: should repository administrators configure which branches are supported for search in the repository settings (or should Gitea administrators decide which branches to configure?)

https://github.com/sourcegraph/zoekt/blob/main/doc/faq.md#can-i-index-multiple-branches

> 2. The search results are not highlighted, it would be nice if this was available.
>    ![image](https://private-user-images.githubusercontent.com/89133723/432998682-5cbbf111-ee06-459d-9b7d-d4b09c523f58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NjY5NzYsIm5iZiI6MTc0NDQ2NjY3NiwicGF0aCI6Ii84OTEzMzcyMy80MzI5OTg2ODItNWNiYmYxMTEtZWUwNi00NTlkLTliN2QtZDRiMDljNTIzZjU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEyVDE0MDQzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4YzYxOGVjMTk3OGNiNjBiOGUzNDMzYzJiYjI2OWQ2YzE3YWIyZWQyZmEyYzhhMzJiOWZhODYxM2IzMzUwYzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Qm7D8vqgiw_2gScaBiKzWpSRJ_ywqhlRKAI893CDQNE)

Yes, the current code highlighting behavior is intended to preserve Gitea's original search style (highlighting from the first matched line in the file through to the last matched line in the file).

Yeah, so it definitely needs to be refactored in the future (Sourcegraph's approach would be better: create a block for each matched line in a file by including the lines above and below it, then combine multiple blocks together, and also implement in-line highlighting).

Furthermore, the entire search interaction needs to be refactored, with Sourcegraph's approach being ideal (supporting exact matching, case sensitivity, match count limits, and distinct regular expression mode).

<img width="871" alt="image" src="https://github.com/user-attachments/assets/7a29b255-dbf6-4815-86e5-998c6f708151" />

I think these are all features that can be supported in the future, but they need to be added in a new Pull Request.

adlternative commented:
@hiifong @lunny @wxiaoguang 

Is anyone willing to continue the review? I have fixed some of the Windows build issues by temporarily avoiding building zoekt on Windows.

wxiaoguang reviewed (commented):
Thank you for inviting me to review. At the moment I don't have a full picture of it (I am not the heavy user of the "code search" feature). So I think the real users (eg: people who up-voted the proposal #33702 or have worked with the code search feature) could help more.

At the moment, my thoughts about this PR:

1. Binary size: it seems that the new dependency increases about 2MB to the output binary (If I remember correctly by the last time I built it). Just FYI.
2. Windows support (https://github.com/sourcegraph/zoekt/pull/909): I don't know whether it means "support" or "not support". But Gitea at the moment still supports Windows.
3. Zoekt syntax: where end users could learn that syntax when using Gitea.
4. Milestone: experimental by a build tag? or include in 1.24 or 1.25?

wxiaoguang commented:
> Ironically this sounded familiar and... yeah it is... #32866 You've hit the same issue as one of the main arguments against embedded otel in gitea :D

Maybe there are differences ..... IMO the "grpc dependency" is a minor factor of the "otel integration". The otel sdk's major concern is that whether it could work out-of-box and provide useful reports by end users, and the official otel sdk just is not able to work with Gitea due to various problems (we have discussed). My suggestion for "otel integration" is that we need to have a general framework & design in Gitea first, make things work, then integrate the otel sdk. So "grpc" is not a real blocker in my mind.

adlternative commented:
> Yeah but it was a larger issue initially at least and one of the blockers and otel does have other issues which I'm working/wasting time on. I just find it funny that it came back randomly like this.
> 
> Size diff in case someone is curious

 Interesting, I wonder what tool was used to generate this size comparison?

adlternative commented:
I tried updating zoekt's google.golang.org/grpc version from v1.69.4 to v1.71.0, but there wasn't much change. I think the core reason is that zoekt extensively uses grpc throughout its codebase, while gitea only uses the grpc status/code package in a few places.

I think this kind of size difference might be unavoidable.

hiifong commented:
> @hiifong @lunny @wxiaoguang
> 
> Is anyone willing to continue the review? I have fixed some of the Windows build issues by temporarily avoiding building zoekt on Windows.

Thank you very much, I will take the time to review the pr

lunny commented:
For introducing zoekt, we may have two choices? One is using zoekt as an embedded one, another is using zoekt as a thirdparty service like elasticsearch? Maybe this PR is the first one and we can introduce the second one once it's stable enough.

hiifong commented:
![image](https://github.com/user-attachments/assets/4b328eb6-547b-4f28-944f-d3792aa8b2c2)

Currently I found a problem, when starting gitea it will index all the repositories at full volume, which causes a great waste of disk space, for gitea.com such a large public website may cause insufficient disk space to the point that the site can not function properly, it would be better if it can do lazy indexing.

adlternative commented:
> ![image](https://private-user-images.githubusercontent.com/89133723/433204914-4b328eb6-547b-4f28-944f-d3792aa8b2c2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2MDY5NDYsIm5iZiI6MTc0NDYwNjY0NiwicGF0aCI6Ii84OTEzMzcyMy80MzMyMDQ5MTQtNGIzMjhlYjYtNTQ3Yi00ZjI4LTk0NGYtZDM3OTJhYThiMmMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE0VDA0NTcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE0NjZhN2ZmNjhiMzIwMTE0ODdjYTFhNjZlZGNmMWFiMThlZWQ1NmVhNmNjMTU1ZTUzYWNlMTA0MmNjNzViNzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lg2M9JgHt5sADlYXTAYB0OUZ5Sx2cBHv2epXZLe4G0M)
> 
> ![image](https://private-user-images.githubusercontent.com/89133723/433206505-1dabcb61-32f5-458f-ad8f-86a619e3aa1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2MDY5NDYsIm5iZiI6MTc0NDYwNjY0NiwicGF0aCI6Ii84OTEzMzcyMy80MzMyMDY1MDUtMWRhYmNiNjEtMzJmNS00NThmLWFkOGYtODZhNjE5ZTNhYTFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE0VDA0NTcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlNmFhYzIyNDJhZjY1M2Y2YjMyMjkzMjRkZWFlNzgwZWNjNDk4OThjMmQyYmE5YWUyZWM3MjkzNmE3NzM2ODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wk4y3KWLnrk_u1WSXn9Nuw927oXHDdjgrWjavjpS1eA)
> 
> Currently I found a problem, when starting gitea it will index all the repositories at full volume, which causes a great waste of disk space, for gitea.com such a large public website may cause insufficient disk space to the point that the site can not function properly, it would be better if it can do lazy indexing.
> 
> I'm hosting my own gitea instance with 2k repositories, which currently take up about 300G of disk space, and I only have 512G of disk space, and 99% of the repositories I don't need to search for, which worries me, maybe it would be better to introduce a third-party service, as @lunny said?

Aha, it sounds like storage space will indeed become an issue. I'll research whether it's possible to break them up through HTTP requests. However, the delayed indexing approach seems simpler, since Gitea's code search only focuses on individual repositories. We'd just need to add a configuration option that prevents Gitea from loading these repository indices at startup, and instead delay loading until the first search is performed...

hiifong commented on modules/indexer/code/zoekt/zoekt.go:
```diff
@@ -0,0 +1,405 @@
+		IsDelta:  true,
+		RepositoryDescription: zoekt.Repository{
+			ID:   uint32(repo.ID),
+			Name: strconv.FormatInt(repo.ID, 10),
```
Would it be better to use {owner}/{repo}?

adlternative replied:
Good advice, I did that. But I'm concerned that when deleting the repository, the index might not be properly deleted because the repository cannot be found.

hiifong commented on options/locale/locale_zh-CN.ini:
```diff
@@ -171,6 +171,8 @@ fuzzy=模糊
 fuzzy_tooltip=包含近似匹配搜索词的结果
 exact=精确
 exact_tooltip=仅包含精确匹配搜索词的结果
+zoekt=Zoekt语法
+zoekt_tooltip=使用 Zoekt 语法进行搜索
```
```suggestion
```
English is sufficient, translations for other languages will be done through Crowdin.

adlternative replied:
OK

hiifong reviewed (approved):
I think it's almost ready, there are still some details we can leave to deal with later, to avoid newcomers waiting too long and lose interest.

wxiaoguang commented:
> Thank you for inviting me to review. At the moment I don't have a full picture of it (I am not the heavy user of the "code search" feature). So I think the real users (eg: people who up-voted the proposal #33702 or have worked with the code search feature) could help more.

Thank you for inviting me to review, I could help to do code-level review, but I am not sure I could speak for real users for this feature (I seldom use code search). So maybe you could invite other maintainers or contributors to help to review.

wxiaoguang commented on modules/indexer/code/zoekt/zoekt.go:
```diff
@@ -0,0 +1,412 @@
+		IsDelta:  true,
+		RepositoryDescription: zoekt.Repository{
+			ID:   uint32(repo.ID),
+			Name: repo.FullName(),
```
Not sure how the `zoekt.Repository` works, but what if a repo is renamed? Then the same repo ID with a different name?

adlternative replied:
It appears that before and after renaming, search and indexing can both proceed normally: everything is associated through the repo ID. However, currently when deleting an index, it first looks up the repo name using the repo ID, and then deletes files on disk that have the repo name prefix. Therefore, there might be some residual data files from before the rename.


<img width="695" alt="image" src="https://github.com/user-attachments/assets/290a4cea-1620-471e-b5c9-928455483d14" />

adl/xx3 -- rename ->  adl/xx422

delete operation will only delete adl%2Fxx422_v16.00000.zoekt...

Perhaps I should check later if there is a way to directly delete an index using the repoid.

wxiaoguang commented on modules/indexer/code/zoekt/zoekt.go:
```diff
@@ -0,0 +1,412 @@
+			return nil, fmt.Errorf("parse regexp keyword %q: %w", keyword, err)
+		}
+
+	case indexer.SearchModeWords:
```
`ZoektSearchModes` doesn't support `SearchModeWords` or `SearchModeExact`, why there are case branches for them?

adlternative replied:
`ZoektSearchModes` contains  `SearchModesExactWords()`, which contains `SearchModeWords` and `SearchModeWords`, 

I think this might be a misunderstanding.

```go
func ZoektSearchModes() []SearchMode {
	return append(SearchModesExactWords(), []SearchMode{
		{
			ModeValue:    SearchModeRegexp,
			TooltipTrKey: "search.regexp_tooltip",
			TitleTrKey:   "search.regexp",
		},
		{
			ModeValue:    SearchModeZoekt,
			TooltipTrKey: "search.zoekt_tooltip",
			TitleTrKey:   "search.zoekt",
		},
	}...)
}
```

lunny commented:
I think this new indexer engine could be merged as an experiment one and stay for some versions to get some feedbacks before it can be become the default one.

hiifong commented:
If I remember correctly, @techknowlogick used to contribute code to zoekt, hopefully he'll have time to review this pr

lunny commented:
Please update app.example.ini to add the newly introduced indexer.

adlternative commented on modules/indexer/code/git.go:
```diff
@@ -37,6 +37,10 @@ func getRepoChanges(ctx context.Context, repo *repo_model.Repository, revision s
 		needGenesis = len(stdout) == 0
 	}
 
+	// TODO: check if zoekt index file meta status is not sync with db index status, if not, get genesis changes
```
Yes, this check can help ensure the correctness of the index data, but it's not necessary at the moment—it can be added in the future.

lunny replied:
But why the below code comments are needed? Maybe they can be removed?

lunny commented:
> @lunny Further review needed. Is there a chance for this PR to be merged into 1.24?

We will release v1.24 very soon. I think this can be merged in v1.25.

wxiaoguang commented:
And some design problems should be addressed (like https://github.com/go-gitea/gitea/pull/33850#discussion_r2046765133, and IIRC there might still a few more, will comment later when I get time). Although it is "experimental", we still need to make the design overall right.

seamon67 commented:
What happened to this? I was looking forward to this one.

kvaster commented:
For us this is also highly awaited feature.

seamon67 commented:
> @kvaster @seamon67 I believe this MR was forgotten in a small corner, and if needed, I can pick it up again.

That would be awesome!

seamon67 commented:
Somebody review this and get this approved please!

lunny commented:
There are still some missing in this PR.
- There’s no version tracking for zoekt similar to what bleve provides. This will make major version upgrades difficult to manage.
- There are still many TODOs and commented-out code left in the PR.
- modules/setting/indexer.go:81 the default indexer path should be changed according to it's bleve or zoekt.
- It seems zoekt needs to be upgrade because it will panic in Go1.25.4

adlternative commented:
> There are still some missing in this PR.
> 
> * There’s no version tracking for zoekt similar to what bleve provides. This will make major version upgrades difficult to manage.

I'll go to the zoekt community to ask if version management can be supported.


> * There are still many TODOs and commented-out code left in the PR.

I think these are all harmless minor TODOs that we could perhaps deploy first and address later.

> * modules/setting/indexer.go:81 the default indexer path should be changed according to it's bleve or zoekt.

Agree, will change.

> * It seems zoekt needs to be upgrade because it will panic in Go1.25.4

I will investigate this issue.

adlternative commented on modules/indexer/code/zoekt/zoekt.go:
```diff
@@ -0,0 +1,413 @@
+		return err
+	}
+
+	repoPathPrefix := repo.OwnerName + "%2F" + repo.Name
```
After thinking about it briefly, this might be related to zoekt's ability to search using the `repo:foo/bar` syntax. However, using repository ID might be better here, as it would be more friendly for deletion/move operations.

adlternative commented:
> > [sourcegraph/zoekt#1000](https://github.com/sourcegraph/zoekt/issues/1000)
> 
> I think we can have a gitea-internal version about the zoekt data format like we did for bleve like this https://github.com/go-gitea/gitea/blob/main/modules/indexer/code/bleve/bleve.go#L75

Oh, I misunderstood what you meant. I thought you were referring to the git tag version, but it turns out you were talking about the data format version.

seamon67 commented:
Don't mind me. I am just glad to see this getting traction.

adlternative commented on modules/indexer/internal/zoekt/indexer.go:
```diff
@@ -0,0 +1,103 @@
+			if err != nil {
+				return false, fmt.Errorf("failed to create index directory: %w", err)
+			}
+		}
```
I think maybe we should also delete the new version of the data?

adlternative commented:
> The comment at https://github.com/go-gitea/gitea/pull/33850/changes#r2638055239 still need to be resolved.

Alright, it looks like this part shouldn't be very important, I've removed this section of comment.

lunny commented:
> > ```
> > # code.gitea.io/gitea/modules/indexer/code/zoekt
> > modules/indexer/code/zoekt/zoekt.go:96:90: undefined: git.WriteCloserError
> > modules/indexer/code/zoekt/zoekt.go:171:24: undefined: git.NewBatch
> > ```
> 
> I don't have any leads right now, at what stage did this occur? During window compile? Or during linting?

Please take a look at the CI error.

adlternative commented:
@lunny fixed, pls check it again!

TheFox0x7 reviewed (commented):
[Query syntax](https://github.com/sourcegraph/zoekt/blob/main/doc/query_syntax.md) documents various search parameters, I've tried them on explore/code endpoint and the handling is inconsistent:
- aliases don't work in general
- case works
- type:repo doesn't return anything despite logs telling `len of (result): 143557`
- branch does nothing as far as I see? 
- repo doesn't seem to work either.

Which of those should and should not work?

adlternative commented on modules/indexer/indexer.go:
```diff
@@ -43,6 +44,21 @@ func SearchModesExactWordsFuzzy() []SearchMode {
+		},
+	}...)
+}
+
```
will fix

adlternative commented:
> > Since Zoekt only indexes code for the default branch by default, `branch` is not supported.
> 
> That sounds like a pretty big shortcoming, ngl...

These are all things that can be optimized later. zoekt most can support up to 64 branches.
https://github.com/sourcegraph/zoekt/blob/main/doc/faq.md#can-i-index-multiple-branches

adlternative commented:
> The search result is still unstable. Try refresh a search page multiple times, the result will change.

I will investigate this intermittent issue.

adlternative commented:
The language filter sidebar was flickering — same search, but the language list would show up in a different order each time. The root cause is that Go's map iteration order is random, so when multiple languages had the same file count, their relative order was essentially a coin flip on every request. Fixed it by adding the language name as a tiebreaker in the sort, so the order is now consistent.
@lunny

adlternative commented:
@lunny please check again, thanks.

lunny commented:
I don't know whether it's a bug of this PR or before. A blank language have been created and chosed.

<img width="1315" height="273" alt="image" src="https://github.com/user-attachments/assets/0c412fdb-b573-4b20-b02b-0248efe48d00" />

adlternative commented:
Files whose language cannot be detected are stored with an empty Language field (""). This caused the language filter sidebar to render a nameless, always-selected button that, when clicked, searched all languages instead of filtering to unclassified files only.

The fix maps the empty language to a visible "Unknown" label, allowing these files to be properly identified and filtered in search results.
<img width="1347" height="617" alt="image" src="https://github.com/user-attachments/assets/329ddfe3-e8e4-4919-8610-3b1a0141b6d1" />

adlternative commented:
@lunny check again, :0

trucpd commented:
up.

adlternative commented:
@lunny The user is getting a bit impatient waiting. Could you please take another look when you have a moment?

seamon67 commented:
up.

bircni commented:
Fixes:
- Build the search query objects directly instead of formatting keywords into `content:"…"` — a keyword containing `"` or `\` no longer errors the search page, and regexp escapes (`\d`, `\b`, `\.`) are no longer stripped
- Treat `GetSkipTake()` as an offset instead of a page index — page 2+ returned the wrong slice, usually empty
- Decide delta vs. full build from `IndexState()` plus a shard cap, falling back to a full rebuild from the tree — shards no longer grow one per push, and changing `MAX_FILE_SIZE` no longer breaks indexing of already-indexed repos permanently
- Apply `opts.SetDefaults()` before the incremental check, so the option hash matches and unchanged repos are actually skipped
- Keep filename matches in `zoekt` search mode, so `file:` queries return results
- Bound searches with `ShardMaxMatchCount` / `MaxDocDisplayCount` instead of loading whole file contents for up to 100k matches per shard
- Call `builder.Finish()` on error paths so temporary shards are cleaned up
- Index binary files by name with empty content (same as bleve) and always consume the blob's trailing newline
- Revert `GetCodeLanguage` to `enry.OtherLanguage`; the `"Unknown"` label broke the repo language bar and spammed a warning on every unclassifiable file view
- Fix the zoekt index metadata init: dead not-exist branch, error returned after recovery, and a bogus "older index with version 0" wipe on every fresh install
- Drop `log.Info` of every user's search query
- Correct the `REPO_INDEXER_INCLUDE`/`EXCLUDE` note in `app.example.ini` — the patterns do apply to zoekt

lunny reviewed (changes requested):
When searching, it will return almost all the lines of a whole file and it's very slow. It should only returned a few lines(i.e. 3 lines).

<img width="1315" height="749" alt="image" src="https://github.com/user-attachments/assets/7822b0f6-ed4a-4065-bab7-1868995cc640" />

hiifong commented:
Maybe we should stay consistent with zoket-webserver and only show matching lines.

<img width="2777" height="3149" alt="image" src="https://github.com/user-attachments/assets/18d59052-15c8-48bf-89a0-0f45fffd6748" />

wxiaoguang commented:
> Maybe we should stay consistent with zoket-webserver and only show matching lines.

And it should be an external service like ElasticSearch, then Gitea only need to call the web server?

hiifong commented:
> > Maybe we should stay consistent with zoket-webserver and only show matching lines.
> 
> And it should be an external service like ElasticSearch, then Gitea only need to call the web server?

no, This is just a simple WebUI search example provided by the official docs.
> ref: https://github.com/sourcegraph/zoekt/tree/main#starting-the-web-server
> Starting the web server
go install github.com/sourcegraph/zoekt/cmd/zoekt-webserver@latest
zoekt-webserver -index ~/.zoekt/
This will start a web server with a simple search UI at http://localhost:6070/. See the [query syntax docs](https://github.com/sourcegraph/zoekt/blob/main/doc/query_syntax.md) for more details on the query language.

wxiaoguang commented:
> > > Maybe we should stay consistent with zoket-webserver and only show matching lines.
> > 
> > 
> > And it should be an external service like ElasticSearch, then Gitea only need to call the web server?
> 
> no, This is just a simple WebUI search example provided by the official docs.

Why not simply use Vibe Coding to build one based on that zoket-webserver to provide "code indexer & search" service?

hiifong commented:
> > > > Maybe we should stay consistent with zoket-webserver and only show matching lines.
> > > 
> > > 
> > > And it should be an external service like ElasticSearch, then Gitea only need to call the web server?
> > 
> > 
> > no, This is just a simple WebUI search example provided by the official docs.
> 
> Why not simply use Vibe Coding to build one based on that zoket-webserver to provide "code indexer & search" service?

Good idea — but let's go by what people actually need.

wxiaoguang commented:
> > > > > Maybe we should stay consistent with zoket-webserver and only show matching lines.
> > > > 
> > > > 
> > > > And it should be an external service like ElasticSearch, then Gitea only need to call the web server?
> > > 
> > > 
> > > no, This is just a simple WebUI search example provided by the official docs.
> > 
> > 
> > Why not simply use Vibe Coding to build one based on that zoket-webserver to provide "code indexer & search" service?
> 
> Good idea — but let's go by what people actually need.

How go to with "cluster" support? https://gitea.com/gitea/helm-gitea/src/branch/main/docs/ha-setup.md

wxiaoguang commented:
Also, the size bloating is not trivial.

* this: 123555282
* main: 112513602

I do not think "embedding everything into one binary" is the right approach, especially when Gitea server still has to run its binary for every git push/pull.

hiifong commented:
> > > > > > Maybe we should stay consistent with zoket-webserver and only show matching lines.
> > > > > 
> > > > > 
> > > > > And it should be an external service like ElasticSearch, then Gitea only need to call the web server?
> > > > 
> > > > 
> > > > no, This is just a simple WebUI search example provided by the official docs.
> > > 
> > > 
> > > Why not simply use Vibe Coding to build one based on that zoket-webserver to provide "code indexer & search" service?
> > 
> > 
> > Good idea — but let's go by what people actually need.
> 
> How go to with "cluster" support? https://gitea.com/gitea/helm-gitea/src/branch/main/docs/ha-setup.md

I think most users run Gitea as a single instance. If they need cluster support, they should choose an appropriate search engine based on their needs.

hiifong commented:
> Also, the size bloating is not trivial.
> 
> * this: 123555282
> * main: 112513602
> 
> I do not think "embedding everything into one binary" is the right approach, especially when Gitea server still has to run its binary for every git push/pull.

I agree with you. Maybe I should split it out as a plugin, or a build tag for it, or do something similar

adlternative commented:
Thanks for the feedback!

1. **Only showing matching lines** (@lunny @hiifong): Agreed — returning nearly the whole file is not ideal. I'll look into how to return only the matching lines with a small amount of context (e.g. ~3 lines), consistent with what zoekt-webserver does.

2. **Running zoekt as an external service** (@wxiaoguang): My concern is that this would be a fairly heavy implementation and quite far from the current state of this PR. Maybe we could treat it as a possible next step / follow-up, rather than blocking this PR on it?

3. **Splitting it out as a plugin or putting it behind a build tag** (@wxiaoguang @hiifong): That sounds reasonable to me as a way to address the binary size bloat. I'll investigate which approach fits better.

adlternative committed:
feat(search): support code search by zoekt

Add zoekt as an optional code search indexer backend
(REPO_INDEXER_TYPE=zoekt). It is guarded by the "zoekt" build tag (on
Unix systems only) so that default builds do not link zoekt at all,
avoiding the >10MB binary size increase; a binary built without the tag
fails fast with a clear error message when configured to use zoekt.

Search results only contain the best matching line of each file with
one context line around it, similar to zoekt-webserver, instead of the
whole file content. The new SearchResult.ContentStartLineNum field
carries the 1-based line number of the first snippet line so the UI
displays correct line numbers; indexers returning whole file content
keep working unchanged.

adlternative commented:
I have updated the PR (rebased on latest main and squashed into a single commit). Both concerns should be addressed now:

1. **Only matching lines are returned now** (@lunny @hiifong): the indexer no longer searches with `Whole: true`. Instead it returns only the best matching line of each file with one context line around it (via zoekt's `NumContextLines`), similar to zoekt-webserver. Line numbers and highlight ranges are computed from the snippet. Searching the git.git source (~4.8k files) now renders 3 lines per result and a full result page takes ~25ms instead of dumping whole files.

2. **The zoekt indexer is now behind a `zoekt` build tag** (@wxiaoguang @hiifong): default builds do not link zoekt at all (`go list -deps` shows zero zoekt packages), so there is no binary size impact unless you opt in. On my machine (darwin/arm64, sqlite build): 148.6MB without the tag vs 162.4MB with it. A binary built without the tag fails fast with a clear error message if `REPO_INDEXER_TYPE = zoekt` is configured.

As for running zoekt as an external service — I'd still prefer to treat that as a possible follow-up rather than blocking this PR, since the build tag already keeps the default binary clean.

hiifong commented:
LGTM
